Add support to ElasticSearch 5.0

lpic10 opened the pull request:
Add support to ElasticSearch 5.0 (currently still in alpha).

Main changes: 
- remove "search_type = count"
- change filtered query and filters to "bool" 

https://www.elastic.co/guide/en/elasticsearch/reference/master/breaking_50_search_changes.html

lpic10 committed:
Add support to ElasticSearch 5.0

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=4970) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=4970) before we can accept your contribution.<br/>

torkelo commented:
this looks like a good start. Some unit tests for the different scenarios would be good very good though.

lpic10 committed:
Add some tests for ES 5.x

bergquist commented:
The commit you branched of from was broken. Please rebase onto the master branch.

lpic10 committed:
Merge branch 'master' of https://github.com/grafana/grafana

atnak commented:
I applied these changes to my current 3.1.0 installation and found that for Annotations to work [this bit here](https://github.com/lpic10/grafana/blob/0ebf469f5f3ec601c9a63dcad40d103cc974fd3d/public/app/plugins/datasource/elasticsearch/datasource.js#L86) also needs a fix.  (I gave the datasource fix a whirl but I mainly only use ES for Annotations.)

These changes were needed to get Annotations working:

At [Line 81](https://github.com/lpic10/grafana/blob/0ebf469f5f3ec601c9a63dcad40d103cc974fd3d/public/app/plugins/datasource/elasticsearch/datasource.js#L81):

```diff
       var queryInterpolated = templateSrv.replace(queryString, {}, 'lucene');
-      var filter = { "bool": { "must": [{ "range": range }] } };
-      var query = { "bool": { "should": [{ "query_string": { "query": queryInterpolated } }] } };
+      var must = [{ "range": range }];
+      var should = [{ "query_string": { "query": queryInterpolated } }];
       var data = {
-        "fields": [timeField, "_source"],
-        "query" : { "filtered": { "query" : query, "filter": filter } },
+        "fields": [ "_source" ],
+        "query" : { "bool": { "must": must, "should": should } },
         "size": 10000
       };
```

At [Line 125](https://github.com/lpic10/grafana/blob/0ebf469f5f3ec601c9a63dcad40d103cc974fd3d/public/app/plugins/datasource/elasticsearch/datasource.js#L125):

```diff
         for (var i = 0; i < hits.length; i++) {
           var source = hits[i]._source;
-          var fields = hits[i].fields;
           var time = source[timeField];

-          if (_.isString(fields[timeField]) || _.isNumber(fields[timeField])) {
-            time = fields[timeField];
-          }
-
           var event = {
```

FWIW, I tried these "bool" queries in ES 1.4 and the same syntax seems to work there also.

lpic10 committed:
Merge branch 'master' of https://github.com/grafana/grafana

lpic10 committed:
Change annotations query for ES 5.x

lpic10 commented:
@atnak: I applied your suggestions with few changes, seems to be working here. I'm also not sure about the fix you mention, thus I've tried to not break older versions behavior.

lpic10 committed:
Change size on terms aggregation for ES 5.x

lpic10 commented:
Added another commit to fix terms aggregation in ES 5.x when size is set to 0, as this is no longer valid. In previous versions, size: 0 in the aggregation was translated to Integer.MAX_VALUE 

My commit will affect queries for older versions as well, but I don't believe anyone is returning that many items in a terms query anyway (in my laptop a 1000 items terms query freezes the browser for 1 minute or so... ). 

Finally I updated the test to reflect this new mandatory behavior on ES 5.x. 

https://www.elastic.co/guide/en/elasticsearch/reference/master/breaking_50_aggregations_changes.html#_literal_size_0_literal_on_terms_significant_terms_and_geohash_grid_aggregations

https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-terms-aggregation.html#_size

torkelo commented:
ok, this looks pretty good thank you for your work on this PR, just needs some testing & merging :)

lpic10 committed:
Fix for annotations query on ES 5.x

lpic10 commented:
Just added another commit due to "fields" parameter being replaced in last alpha release.

https://www.elastic.co/guide/en/elasticsearch/reference/master/breaking_50_search_changes.html#_literal_fields_literal_parameter

atnak commented on public/app/plugins/datasource/elasticsearch/datasource.js:
```diff
@@ -124,11 +128,12 @@ function (angular, _, moment, kbn, ElasticQueryBuilder, IndexPattern, ElasticRes
-
-          if (_.isString(fields[timeField]) || _.isNumber(fields[timeField])) {
-            time = fields[timeField];
+          if (typeof hits[i].fields !== 'undefined') {
```
Should this be `if (this.esVersion < 5) {` for consistency since the above `// fields field not supported on ES 5.x` change means we're no longer requesting fields ES >= 5?

atnak commented on public/app/plugins/datasource/elasticsearch/query_builder.js:
```diff
@@ -199,29 +228,48 @@ function (queryDef) {
-            "bool": {
-              "must": [{"range": this.getRangeFilter()}]
+    var query = {};
+    if (this.esVersion >= 5) {
```
Unless I'm missing something important, this branching may be unnecessary in favor of just adopting `bool` queries all out since:

* [`filter` queries has been running internally as `bool` queries since ES 2.0](https://www.elastic.co/blog/better-query-execution-coming-elasticsearch-2-0) so the condition should probably be at least `this.esVersion >= 2`.  Otherwise, if this is feared to cause a regression for ES 2, then ES 5 support is going to be a broken too.

* The `bool` query [appears to have been supported since 1.3](https://www.elastic.co/guide/en/elasticsearch/reference/1.3/query-dsl-bool-query.html) and [even 0.90](https://www.elastic.co/guide/en/elasticsearch/reference/0.90/query-dsl-bool-query.html).  I tried `bool` queries with ES 1.4 locally and it doesn't seem to barf.

* There's no similar branching for Annotations so if we're going to assume branching is required, Annotations will be a regression for ES < 5.

lpic10 replied:
Yes, maybe that branching is unnecessary, I wasn't able to test in previous ES versions so it was my intention to not touch code there. 

The annotations change was obvious enough for me to understand the implications in other versions, branching could be an option though.

lpic10 replied:
At the time I did very very basic tests using this 5.x query construction in ES 2.4 and it seemed to work fine.
What kind of issue do you see with `analyze_wildcard = true` in 2.x ? 

I can't think of a particular case where someone needs `analyze_wildcard = true` in grafana but I also don't see any particular reason to change.

lpic10 committed:
Merge branch 'master' of github.com:grafana/grafana

lpic10 committed:
Merge branch 'master' of https://github.com/grafana/grafana

lpic10 commented:
Updated to latest grafana + some fixes on tests. Tested on ElasticSearch 5.0-beta1

lpic10 commented:
linking to #5740

kerk1v commented:
Hello, 
I have tested this in my local install, works flawlessly but it looks like "nested" templating is broken. 

Let me explain, before I had a dashboard with two variables: 
- $prov	{"find": "terms", "field": "province"}
- $loc	{"find": "terms", "field": "location", "query": "province: $prov"}

so when I selected a province in $prov variable, it would only show $loc values actually located in $prov. Now regardless of the value of $prov it always shows all possible values of $loc. Sort of impractical, since there is a list of more than 800 possible values for $loc :(

Also, I don't know if this has broken due to the latest grafana snapshot (4.0.0-pre1) or your patch (which I have manually applied upon said grafana/master and then built myself), but thought it would be cautious to let you know as well. I will also open an issue in the grafana repo

lpic10 committed:
Fix templating using query

lpic10 committed:
Merge branch 'master' of https://github.com/grafana/grafana

lpic10 commented:
Hello @kerk1v , that should be fixed now, thanks

megakid commented:
@lpic10 I have this running from source with our Elastic 5.0 instance as the primary data source.  I noticed today when setting up a new "Raw Document" Table panel that I got this response: 

    {
      "error": {
        "root_cause": [
          {
            "type": "parsing_exception",
            "reason": "The field [fields] is no longer supported, please use [stored_fields] to retrieve stored fields or _source filtering if the field is not stored",
            "line": 1,
            "col": 265
          }
        ],
        "type": "parsing_exception",
        "reason": "The field [fields] is no longer supported, please use [stored_fields] to retrieve stored fields or _source filtering if the field is not stored",
        "line": 1,
        "col": 265
      },
      "status": 400
    }

Request details:

Url	/grafana/api/datasources/proxy/1/_msearch
Method	POST
Accept	application/json, text/plain, */*
Content-Type	application/json;charset=utf-8

No request parameters listed.

megakid commented:
@lpic10 I also came across another issue when trying to display a term bucketed document count graph.  I cloned it from the same graph but averaging a metric instead of counting it.  When changing the Metric type from Avg to Count it responded with this:

    {
      "error": {
        "root_cause": [
          {
            "type": "parsing_exception",
            "reason": "Could not find aggregator type [count] in [1]",
            "line": 1,
            "col": 388
          }
        ],
        "type": "parsing_exception",
        "reason": "Could not find aggregator type [count] in [1]",
        "line": 1,
        "col": 388
      },
      "status": 400
    }

Request details

Url	/grafana/api/datasources/proxy/1/_msearch
Method	POST
Accept	application/json, text/plain, */*
Content-Type	application/json;charset=utf-8

`{"search_type":"query_then_fetch","ignore_unavailable":true,"index":["myindex-2016.11.06"]}
{"size":0,"query":{"bool":{"must":[{"range":{"@timestamp":{"gte":"1478461200000","lte":"1478461500000","format":"epoch_millis"}}},{"query_string":{"analyze_wildcard":true,"query":"(Env: \"Production\" OR instanceName: \"Production\") AND _type: ResponseInfo"}}]}},"aggs":{"3":{"terms":{"field":"ResponseInfo.Url","size":20,"order":{"1":"desc"}},"aggs":{"1":{"count":{"field":"ResponseInfo.Duration"}},"2":{"date_histogram":{"interval":"200ms","field":"@timestamp","min_doc_count":0,"extended_bounds":{"min":"1478461200000","max":"1478461500000"},"format":"epoch_millis"},"aggs":{}}}}}}`

lpic10 commented:
Hello @megakid ,
For the first issue I will send a patch later today, but I cannot reproduce the last one you reported (or I didn't understand well what you are doing). I think a screenshot of the query editor may help.

lpic10 committed:
Merge branch 'master' of https://github.com/grafana/grafana

lpic10 committed:
Fix raw document query on ES 5.x

funkomatic commented:
Hi, 
I tested with the last build from master branch (https://grafanarel.s3.amazonaws.com/builds/grafana-latest.linux-x64.tar.gz) and I still get the following error : 
`{
  "error": {
    "root_cause": [
      {
        "type": "illegal_argument_exception",
        "reason": "No search type for [count]"
      }
    ],
    "type": "illegal_argument_exception",
    "reason": "No search type for [count]"
  },
  "status": 400
}
			`
Any suggestions ? 
Thanks,

lpic10 commented:
Hi @funkomatic , this PR is not merged yet into master. You can apply this as a patch on top of master and build grafana yourself for now.

funkomatic commented:
Thanks for the info, however as I have many issues with the build on my station, I'll wait for the merge & build.

ameade commented:
Hey merging this would really save me a lot of trouble. I will currently build this from source and post here if I find anything wrong.

torkelo commented:
we will try to find time to merge this soon

txisme commented:
Hey, where can i download grafana with ES 5.0 support?

bergquist commented:
@txisme Right now you have to build this branch by yourself. Support for 5.0 will be added to 4.1.0

Petyas commented:
@bergquist On when is please version 4.1.0 planned? Because currently was released version 4.0 Beta 2, than will be 4.0 stable a than, god knows, 4.1. It's not possible only build datasource to version 4.0B2? I need at least basic support for ES 5.x.. Thanks a lot for info.

txisme commented:
Last time i checked, Grafana with support for ES 5.0 will be out on the 19th of December

phutchins commented:
Didn't realize we didn't have support for ES 5.0 yet and upgraded. Unfortunately my dashboards are all broken now :(. Looking forward to getting this merged!

funkomatic commented:
Sorry maybe I did not get it. So ES 5.0 support will be _released_ in 4.1, around 19th December, but I suppose the code will be pushed/merged to master  before this date and that we will be able to download the build from master branch correct ?

bergquist commented:
@funkomatic your correct

txisme commented:
So is there any estimation for when master will have ES 5.0 support?

jpyth commented:
For those waiting as I was(Made the mistake of upgrading without confirming grafana supported --woops). 

It was fairly straight forward to build lpic10's branch and run it following the steps in the README. Which so far has worked great with ES 5.0.2 .

No not production safe; but in my case was better than nothing and will suffice till the master branch is available.

That said; eager to see this in the official release:)

funkomatic commented:
On my side I managed, after a short struggle with go 1.7 install, to build it from lpic10's branch and it works pretty well.
Just a (out of topic) question : is there a way of building the distribution zip from the source and what would it be ?

lpic10 commented:
I've been building since few months back the rpm package for my system using "go run build.go package" when merging grafana's master branch from time to time (weekly usually). With that I can upgrade or downgrade between my own versions and official versions easily.

bergquist commented:
The code seems fine but I would really like to avoid code branching depending on ES version when possible. 

Reducing code complexity is really important for open source software since so many different people will be contributing to the code base over a long period of time. The elastic datasource is kinda complex as it is :)

bergquist reviewed (commented):
The code seems fine but I would really like to avoid code branching depending on ES version when possible.

Reducing code complexity is really important for open source software since so many different people will be contributing to the code base over a long period of time. The elastic datasource is kinda complex as it is :)

txisme commented:
Why not do a new release of Grafana only supporting ES 5.0
For those that are using previous versions of ES, they can always use previous versions of Grafana.

lpic10 commented:
For info, testing Elasticsearch 5.1.1 in here with this, it seems to be working fine

bergquist commented:
Merged! Will be available in 4.1.0

Made some minor changes
- Removed support for elastic 1.x
- Removed some code branching to reduce code complexity

bergquist commented:
Available for download at http://grafana.org/builds/

phutchins commented:
Wonderful news! Thank you!!
